### PR TITLE
Dev thread safe pwm

### PR DESCRIPTION
**Pull request by sjpark608**
theading lock is implemented little differently in PWM module. This is due to how the difference in operation. Unlike other peripherals, PWM file has to be open for the duration of its usage. The init_pwm() method already has checking for whether it's already open or not by using __pwm_dev dictionary defined as class variable. However, close() and set_config method require a locking.

- This PR implements the thread lock for both PWM1 and PWM2. 
- Moved __pwm_dev dictionary to class variable

### CODE REVIEW

File: src/test_edgepi/unit_tests/test_pwm/test_edgepi_pwm.py
Comment (farzadpanahi) [inline src/test_edgepi/unit_tests/test_pwm/test_edgepi_pwm.py:1]: can we have specific tests for checking multi threading, similar to what we have for other peripherals?

File: src/edgepi/pwm/edgepi_pwm.py [hunk 1]
```diff
@@ -175,18 +180,20 @@ def close(self, pwm_num: PWMPins):
...
         """
-        if pwm_num is None or pwm_num not in self.__pwm_devs:
-            raise ValueError(f"get_enabled: PWM number is missing {pwm_num}")
-        self.__pwm_devs[pwm_num].close_pwm()
-        self.__pwm_devs[pwm_num] = None
+        if pwm_num is None or pwm_num not in EdgePiPWM.__pwm_devs:
+            raise ValueError(f"close: PWM number is missing {pwm_num}")
+        with EdgePiPWM.__lock_pwm[pwm_num]:
+            EdgePiPWM.__pwm_devs[pwm_num].close_pwm()
+            EdgePiPWM.__pwm_devs[pwm_num] = None
```
Comment (farzadpanahi) [inline src/edgepi/pwm/edgepi_pwm.py:187]: if `EdgePiPWM.__pwm_devs[pwm_num] = None` has to be done on close, I think we should wrap `close_pwm` in try-catch-finally as well. something like:

```
        with EdgePiPWM.__lock_pwm[pwm_num]:
            try:
                EdgePiPWM.__pwm_devs[pwm_num].close_pwm()
            except:
                raise
            finally:
                 EdgePiPWM.__pwm_devs[pwm_num] = None
```

**Review by farzadpanahi — CHANGES_REQUESTED** [review]

File: src/edgepi/pwm/edgepi_pwm.py (hunk 1 above)
Comment (sjpark608) [reply src/edgepi/pwm/edgepi_pwm.py]: [7725deb](https://github.com/EdgePi-Cloud/edgepi-python-sdk/pull/400/commits/7725deb4c20419d13af8a8e01cf95da3c8105d49)

**Review by farzadpanahi — APPROVED** [review]

File: src/test_edgepi/unit_tests/test_pwm/test_edgepi_pwm.py
Comment (sjpark608) [reply src/test_edgepi/unit_tests/test_pwm/test_edgepi_pwm.py:1]: [7725deb](https://github.com/EdgePi-Cloud/edgepi-python-sdk/pull/400/commits/7725deb4c20419d13af8a8e01cf95da3c8105d49)

the target test function includes three lines

- init_pwm (instantiate pwm device file)
- set_config (write to device file)
- close (close the device file)

There are 100 threads generated with the target function for each iteration. The test is failing due to the each some threads are trying to write while in other thread the files are close. I should probably implement some checks inside get/set config functions whether the file is still open or not.

Comment (sjpark608) [reply src/test_edgepi/unit_tests/test_pwm/test_edgepi_pwm.py:1]: removing the close() in the target function makes the test pass

**Comment by sjpark608** [comment]
We've discussed what'd be the expected behaviour of reading/writing closed file. The program should bubble up the exception thrown when read/write the closed file to notify there's a thread closed the PWM file (intentionally and unintentionally).

**Pull request merged**